Menus/PieMenuGUI.cpp

```diff
@@ -1,1416 +1,634 @@
+	}
 
-		x += width / 2 + gap * 2;
-		
-		// Draw right part of the list
-		for (int i = 0; i < half; ++i)
-		{
-			BITMAP * pIcon = pInventory->at(i)->GetGraphicalIcon();
-			//rect(pTargetBitmap, x - gap, y - gap, x + pIcon->w + gap, y + height + gap, 5);
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool PieMenuGUI::HandleNonMouseInput() {
+		std::map<ControlState, PieSlice *> controlStateSlices = {
+			{PRESS_UP, &m_UpSlice},
+			{PRESS_LEFT, &m_LeftSlice},
+			{PRESS_DOWN, &m_DownSlice},
+			{PRESS_RIGHT, &m_RightSlice}
+		};
+		std::map<ControlState, std::pair<std::pair<float, float>, bool>> controlStateSelectionZoomAngles = {
+			{PRESS_UP, {{m_LeftSlice.GetMidAngle(), m_RightSlice.GetMidAngle()}, true}},
+			{PRESS_LEFT, {{m_DownSlice.GetMidAngle(), m_UpSlice.GetMidAngle()}, false}},
+			{PRESS_DOWN, {{m_RightSlice.GetMidAngle(), m_LeftSlice.GetMidAngle()}, false}},
+			{PRESS_RIGHT, {{m_UpSlice.GetMidAngle(), m_DownSlice.GetMidAngle()}, true}}
+		};
+		std::map<ControlState, std::pair<std::pair<float, float>, bool>> controlStateSelectionStepAngles = {
+			{PRESS_UP, {{m_UpSlice.GetMidAngle(), m_DownSlice.GetMidAngle()}, true}},
+			{PRESS_LEFT, {{m_LeftSlice.GetMidAngle(), 0}, false}},
+			{PRESS_DOWN, {{m_DownSlice.GetMidAngle(), m_UpSlice.GetMidAngle()}, false}},
+			{PRESS_RIGHT, {{0, m_LeftSlice.GetMidAngle()}, true}}
+		};
+
+		for (const auto &[controlState, pieSlice] : controlStateSlices) {
+			if (m_MenuController->IsState(controlState)) {
+				const auto &[firstSelectionZoomAngle, secondSelectionZoomAngle] = controlStateSelectionZoomAngles.at(controlState).first;
+				bool shouldZoom = !m_HoveredSlice;
+				shouldZoom |= controlStateSelectionZoomAngles.at(controlState).second ?
+					(m_CursorAngle > firstSelectionZoomAngle && m_CursorAngle < secondSelectionZoomAngle) :
+					(m_CursorAngle > firstSelectionZoomAngle || m_CursorAngle < secondSelectionZoomAngle);
+
+				if (shouldZoom) {
+					return SelectPieSlice(pieSlice, true);
+				} else if (m_HoveredSlice != pieSlice) {
+					const auto &[firstSelectionStepAngle, secondSelectionStepAngle] = controlStateSelectionStepAngles.at(controlState).first;
+					int stepDirection = controlStateSelectionStepAngles.at(controlState).second ?
+						((m_CursorAngle > firstSelectionStepAngle && m_CursorAngle < secondSelectionStepAngle) ? -1 : 1) :
+						((m_CursorAngle > firstSelectionStepAngle || m_CursorAngle < secondSelectionStepAngle) ? -1 : 1);
+
+					vector<PieSlice *>::iterator sliceToSelect = std::find(m_CurrentSlices.begin(), m_CurrentSlices.end(), m_HoveredSlice) + stepDirection;
+					if (sliceToSelect >= m_CurrentSlices.end()) {
+						sliceToSelect = m_CurrentSlices.begin();
+					} else if (sliceToSelect < m_CurrentSlices.begin()) {
+						sliceToSelect = m_CurrentSlices.end() - 1;
+					}
+
+					if ((*sliceToSelect) != nullptr) {
+						return SelectPieSlice((*sliceToSelect), true);
```

Comment:
Hmm, I don't really see how this could have crashed? The lines above it basically ensure sliceToSelect is set in bounds, if it's >= slices end then it gets set to the beginning, and if it's < the beginning it gets set to end - 1. This should be safe, right?

Anyway, I haven't been able to reproduce this. If you can get more details, that'd be super helpful.

Reply:
I dunno, maybe you somehow fixed it. ¯\\\_(ツ)_/¯